src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl

```diff
@@ -75,3 +78,144 @@ end
+    η = free_surface.η
+    η = η_save
+    fill_halo_regions!(RHS   , η.boundary_conditions, model.architecture, model.grid)
+    fill_halo_regions!(η.data, η.boundary_conditions, model.architecture, model.grid)
+    ##  need to subtract Azᵃᵃᵃ(i, j, 1, grid)*η[i,j, 1]/(g*Δt^2)
+    event = add_previous_free_surface_contribution(free_surface, model, Δt )
+    wait(device(model.architecture), event)
+    fill_halo_regions!(RHS   , η.boundary_conditions, model.architecture, model.grid)
+    ## RHS .= RHS .+ free_surface.η.data/Δt
+
+    ## Then we can invoke solve_for_pressure! on the right type via calculate_pressure_correction!
+    x  = free_surface.implicit_step_solver.solver.settings.x
+    x .= η.data
+    fill_halo_regions!(x ,η.boundary_conditions, model.architecture, model.grid)
+    solve_poisson_equation!(free_surface.implicit_step_solver.solver, RHS, x)
+    ## exit()
+    fill_halo_regions!(x ,η.boundary_conditions, model.architecture, model.grid)
+    free_surface.η.data .= x
+
+    ## Once we have η we can update u* and v* with pressure gradient just as in pressure_correct_velocities!
+
+    ## The explicit form of this function defaults to returning an event, we do the same for now.
+    return event
+end
+
+
+using Oceananigans.Architectures: device
+using Oceananigans.Operators: ΔzC
+
+"""
+Compute the vertical integrated transport from the bottom to z=0 (i.e. linear free-surface)
+
+    `U^{*} = ∫ [(u^{*})] dz`
+    `V^{*} = ∫ [(v^{*})] dz`
+"""
+### Note - what we really want is RHS = divergence of the vertically integrated transport
+###        we can optimize this a bit later to do this all in one go to save using intermediate variables.
+function compute_vertically_integrated_transport!(free_surface, model)
+
+    event = launch!(model.architecture,
+                    model.grid,
+                    :xy,
+                    _compute_vertically_integrated_transport!,
+                    model.velocities,
+                    model.grid,
+                    free_surface.barotropic_transport,
+                    dependencies=Event(device(model.architecture)))
+
+    return event
+end
+
+@kernel function _compute_vertically_integrated_transport!(U, grid, barotropic_transport )
+    i, j = @index(Global, NTuple)
+    # U.w[i, j, 1] = 0 is enforced via halo regions.
+    barotropic_transport.u[i, j, 1] = 0.
+    barotropic_transport.v[i, j, 1] = 0.
+    @unroll for k in 1:grid.Nz
+        #### @inbounds barotropic_transport.u[i, j, 1] += U.u[i, j, k-1]*Δyᶠᶜᵃ(i, j, k, grid)*Δzᵃᵃᶜ(i, j, k, grid)
+        #### @inbounds barotropic_transport.v[i, j, 1] += U.v[i, j, k-1]*Δyᶠᶜᵃ(i, j, k, grid)*Δzᵃᵃᶜ(i, j, k, grid)
+        @inbounds barotropic_transport.u[i, j, 1] += U.u[i, j, k]*Δyᶠᶜᵃ(i, j, k, grid)*ΔzC(i, j, k, grid)
+        @inbounds barotropic_transport.v[i, j, 1] += U.v[i, j, k]*Δxᶜᶠᵃ(i, j, k, grid)*ΔzC(i, j, k, grid)
```

Comment:
Since we multiple by face areas, what we are calling `barotropic_transport` is actually the `barotropic_mass_flux`, right?

Reply:
Yes

Reply:
I guess strictly its a volume flux, so I went with that.